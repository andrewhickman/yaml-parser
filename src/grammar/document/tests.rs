use insta::assert_yaml_snapshot;

use crate::{cursor::Cursor, grammar::tests::parse, Diagnostic, Receiver};

use super::Document;

fn prefix<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<Document<'s>, Diagnostic> {
    super::prefix(cursor, receiver, true)
}

fn prefix_unterminated<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized + 'static),
) -> Result<Document<'s>, Diagnostic> {
    super::prefix(cursor, receiver, false)
}

#[test]
fn test_prefix() {
    assert_yaml_snapshot!(parse(prefix, ""));
    assert_yaml_snapshot!(parse(prefix, "\u{feff}"));
    assert_yaml_snapshot!(parse(prefix, " "));
    assert_yaml_snapshot!(parse(prefix, "\u{feff} "));
    assert_yaml_snapshot!(parse(prefix, "# comment"));
    assert_yaml_snapshot!(parse(prefix, "\u{feff}# comment"));
    assert_yaml_snapshot!(parse(prefix, "\u{feff}  # comment\n\n#comment2\n"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2 # yaml version"));
    assert_yaml_snapshot!(parse(
        prefix,
        "%TAG !yaml! tag:yaml.org,2002: # tag\n# directive"
    ));
    assert_yaml_snapshot!(parse(prefix, "%FOO bar baz"));
    assert_yaml_snapshot!(parse(prefix, "%FOO bar baz\n# reserved"));
    assert_yaml_snapshot!(parse(
        prefix,
        "%YAML 1.2\n%TAG !yaml! tag:yaml.org,2002:\r\n%FOO bar baz"
    ));
    assert_yaml_snapshot!(parse(
        prefix,
        "%YAML 1.2\n# yaml\n%TAG !yaml! tag:yaml.org,2002:\r\n\t#tag\n%FOO bar baz"
    ));
    assert_yaml_snapshot!(parse(prefix, "\u{feff}%YAML 1.2"));
    assert_yaml_snapshot!(parse(prefix, "\u{feff}%TAG !yaml! tag:yaml.org,2002:"));
    assert_yaml_snapshot!(parse(prefix, "\u{feff}%FOO bar baz"));
    assert_yaml_snapshot!(parse(
        prefix,
        "\u{feff}%YAML 1.2\n%TAG !yaml! tag:yaml.org,2002:\r\n%FOO bar baz"
    ));
    assert_yaml_snapshot!(parse(
        prefix,
        "\u{feff}%YAML 1.2\n# yaml\n%TAG !yaml! tag:yaml.org,2002:\r\n\t#tag\n%FOO bar baz"
    ));
}

#[test]
fn test_yaml_directive() {
    assert_yaml_snapshot!(parse(prefix, "%YAML"));
    assert_yaml_snapshot!(parse(prefix, "%YAML "));
    assert_yaml_snapshot!(parse(prefix, "%YAML\n"));
    assert_yaml_snapshot!(parse(prefix, "%YAML foo"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 0"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 0\n"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 0foo"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 0."));
    assert_yaml_snapshot!(parse(prefix, "%YAML 0.\n"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 0.foo"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 0.0foo"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2 foo"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2 \n"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2\t\n"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2\n"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2\r\n"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 0.0"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.0"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.1"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 9999999999999.1"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.9999999999999"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2 \n%YAML 1.2"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.3"));
    assert_yaml_snapshot!(parse(prefix, "%YAML 1.2\n---\nfoo"));
}

#[test]
fn test_tag_directive() {
    assert_yaml_snapshot!(parse(prefix, "%TAG"));
}
