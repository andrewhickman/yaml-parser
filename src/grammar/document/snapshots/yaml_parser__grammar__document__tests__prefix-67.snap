---
source: src/grammar/document/tests.rs
expression: "parse(prefix,\n    \"%YAML invalid\\n# yaml\\n%TAG !yaml! tag:yaml.org,2002:\\n%FOO bar baz\\n---\")"
---
- token: directive
  stream: "%"
  span:
    start:
      index: 0
      line: 0
      column: 0
    end:
      index: 1
      line: 0
      column: 1
- token: directive_name
  stream: YAML
  span:
    start:
      index: 1
      line: 0
      column: 1
    end:
      index: 5
      line: 0
      column: 5
- token: separator
  stream: " "
  span:
    start:
      index: 5
      line: 0
      column: 5
    end:
      index: 6
      line: 0
      column: 6
- token: error
  stream: invalid
  span:
    start:
      index: 6
      line: 0
      column: 6
    end:
      index: 13
      line: 0
      column: 13
- severity: error
  message: "expected a decimal digit (0-9), but found 'i'"
  span:
    start:
      index: 6
      line: 0
      column: 6
    end:
      index: 7
      line: 0
      column: 7
  stream: i
- token: break
  stream: "\n"
  span:
    start:
      index: 13
      line: 0
      column: 13
    end:
      index: 14
      line: 1
      column: 0
- token: comment
  stream: "# yaml"
  span:
    start:
      index: 14
      line: 1
      column: 0
    end:
      index: 20
      line: 1
      column: 6
- token: break
  stream: "\n"
  span:
    start:
      index: 20
      line: 1
      column: 6
    end:
      index: 21
      line: 2
      column: 0
- token: directive
  stream: "%"
  span:
    start:
      index: 21
      line: 2
      column: 0
    end:
      index: 22
      line: 2
      column: 1
- token: directive_name
  stream: TAG
  span:
    start:
      index: 22
      line: 2
      column: 1
    end:
      index: 25
      line: 2
      column: 4
- token: separator
  stream: " "
  span:
    start:
      index: 25
      line: 2
      column: 4
    end:
      index: 26
      line: 2
      column: 5
- token: tag_handle
  stream: "!yaml!"
  span:
    start:
      index: 26
      line: 2
      column: 5
    end:
      index: 32
      line: 2
      column: 11
- token: separator
  stream: " "
  span:
    start:
      index: 32
      line: 2
      column: 11
    end:
      index: 33
      line: 2
      column: 12
- token: tag_prefix
  stream: "tag:yaml.org,2002:"
  span:
    start:
      index: 33
      line: 2
      column: 12
    end:
      index: 51
      line: 2
      column: 30
- token: break
  stream: "\n"
  span:
    start:
      index: 51
      line: 2
      column: 30
    end:
      index: 52
      line: 3
      column: 0
- token: directive
  stream: "%"
  span:
    start:
      index: 52
      line: 3
      column: 0
    end:
      index: 53
      line: 3
      column: 1
- token: directive_name
  stream: FOO
  span:
    start:
      index: 53
      line: 3
      column: 1
    end:
      index: 56
      line: 3
      column: 4
- severity: warning
  message: "unknown directive 'FOO'"
  span:
    start:
      index: 53
      line: 3
      column: 1
    end:
      index: 56
      line: 3
      column: 4
  stream: FOO
- token: separator
  stream: " "
  span:
    start:
      index: 56
      line: 3
      column: 4
    end:
      index: 57
      line: 3
      column: 5
- token: directive_parameter
  stream: bar
  span:
    start:
      index: 57
      line: 3
      column: 5
    end:
      index: 60
      line: 3
      column: 8
- token: separator
  stream: " "
  span:
    start:
      index: 60
      line: 3
      column: 8
    end:
      index: 61
      line: 3
      column: 9
- token: directive_parameter
  stream: baz
  span:
    start:
      index: 61
      line: 3
      column: 9
    end:
      index: 64
      line: 3
      column: 12
- token: break
  stream: "\n"
  span:
    start:
      index: 64
      line: 3
      column: 12
    end:
      index: 65
      line: 4
      column: 0
- token: directives_end
  stream: "---"
  span:
    start:
      index: 65
      line: 4
      column: 0
    end:
      index: 68
      line: 4
      column: 3
- value:
    explicit: true
    version: ~
    tags:
      "!yaml!": "tag:yaml.org,2002:"

